src/gluonts/torch/model/deepar/deepar_network.py

```diff
@@ -0,0 +1,430 @@
+
+    @staticmethod
+    def get_lagged_subsequences(
+        sequence: torch.Tensor,
+        indices: List[int],
+        subsequences_length: int = 1,
+    ) -> torch.Tensor:
+        """
+        Returns lagged subsequences of a given sequence.
+
+        Parameters
+        ----------
+        sequence : Tensor
+            the sequence from which lagged subsequences should be extracted.
+            Shape: (N, T, C).
+        indices : List[int]
+            list of lag indices to be used.
+        subsequences_length : int
+            length of the subsequences to be extracted.
+
+        Returns
+        --------
+        lagged : Tensor
+            a tensor of shape (N, S, C, I), where S = subsequences_length and
+            I = len(indices), containing lagged subsequences. Specifically,
+            lagged[i, j, :, k] = sequence[i, -indices[k]-S+j, :].
+        """
+        sequence_length = sequence.shape[1]
+
+        assert max(indices) + subsequences_length <= sequence_length, (
+            f"lags cannot go further than history length, found lag {max(indices)} "
+            f"while history length is only {sequence_length}"
+        )
+        assert all(lag_index >= 0 for lag_index in indices)
+
+        lagged_values = []
+        for lag_index in indices:
+            begin_index = -lag_index - subsequences_length
+            end_index = -lag_index if lag_index > 0 else None
+            lagged_values.append(sequence[:, begin_index:end_index, ...])
+        return torch.stack(lagged_values, dim=-1)
+
+    def unroll_encoder(
+        self,
+        feat_static_cat: torch.Tensor,
+        feat_static_real: torch.Tensor,
+        past_time_feat: torch.Tensor,
+        past_target: torch.Tensor,
+        past_observed_values: torch.Tensor,
+        future_time_feat: Optional[torch.Tensor] = None,
+        future_target: Optional[torch.Tensor] = None,
+    ) -> Tuple[
+        torch.Tensor, Union[torch.Tensor, List], torch.Tensor, torch.Tensor
+    ]:
+        if not self.training:
+            time_feat = past_time_feat[:, -self.context_length :, ...]
+            sequence = past_target
+            subsequences_length = self.context_length
+        else:
+            time_feat = torch.cat(
+                (
+                    past_time_feat[:, -self.context_length :, ...],
+                    future_time_feat,
+                ),
+                dim=1,
+            )
+            sequence = torch.cat((past_target, future_target), dim=1)
+            subsequences_length = self.context_length + self.prediction_length
+
+        lags = self.get_lagged_subsequences(
+            sequence=sequence,
+            indices=self.lags_seq,
+            subsequences_length=subsequences_length,
+        )
+
+        # scale is computed on the context length last units of the past target
+        # scale shape is (batch_size, 1, *target_shape)
+        _, scale = self.scaler(
+            past_target[:, self.context_length :, ...],
```

Comment:
Shouldn't this be `past_target[:, -self.context_length :, ...]`? Otherwise it starts with the `context_length` index until the end of that axis? Same for observed values below.

Reply:
yes also my fault! good catch!